EIPS/eip-5496.md

```diff
@@ -0,0 +1,205 @@
+
+    /// @notice Check if a user has a certain privilege
+    /// @param tokenId The identifier of the queried NFT
+    /// @param privilegeId The identifier of the queried privilege
+    /// @param user The address of the queried user
+    /// @return Whether a user has a certain privilege
+    function hasPrivilege(uint256 tokenId, uint256 privilegeId, address user) external view returns(bool);
+}
+```
+Every contract implementing the `ERC5496` SHOULD set a maximum privilege number before setting any privilege, the `privilegeId` MUST NOT be greater than the maximum privilege number.
+
+The `PrivilegeAssigned` event MUST be emitted when `setPrivilege` is called.
+
+The `PrivilegeTransfered` event MUST be emitted when `transferPrivilege` is called.
+
+The `supportsInterface` method MUST return `true` when called with `0x076e1bbb`.
+
+```solidity
+/// @title Cloneable extension - Optional for ERC-721
+interface IERC721Cloneable {
+    /// @notice Emitted when set the `privilege ` of a NFT cloneable.
+    event PrivilegeCloned(uint tokenId, uint privId, address from, address to);
+
+    /// @notice set a certain privilege cloneable
+    /// @param tokenId The identifier of the queried NFT
+    /// @param privilegeId The identifier of the queried privilege
+    /// @param referrer The address of the referrer
+    /// @return Whether the operation is sucessful or not
+    function clonePrivilege(uint tokenId, uint privId, address referrer) external returns (bool);
+}
+```
+The `PrivilegeCloned` event MUST be emitted when `clonePrivilege` is called.
+
+## Rationale
+
+### Multiple privileges
+
+EIP-721 only records the ownership and its transfer, the privileges of an NFT are not recorded on-chain. This extension would allow merchants/projects to give out a certain privilege to a specified group of people, and owners of the privileges can manage each one of the privileges independently. This facilitates a great possibility for NFTs to have real usefulness. 
+
+### Sharable privileges
+
+The number of privilege holders is limited by the number of NFTs if privileges are non-sharable. A sharable privilege means the original privilege holder can copy the privilege and give it to the others, not transferring his/her own privilege to the others. This mechanism greatly enhances the spread of privileges as well as the adoption of NFTs. 
+
+## Backwards Compatibility
+
+ERC5496 is compatible with any kind of NFTs that follow the EIP-721 standard. It only adds more functions and data structures without interfering with the original EIP-721 standard.
+
+## Test Cases
+Test cases are implemented in the reference implementation
+repository [here](https://github.com/wnft/ERC5496/tree/main/test)
+
+## Reference Implementation
+The following is a snippet for reference implementation of the `ERC5496` extension. The full repository can be
+found [here](https://github.com/wnft/ERC5496/blob/main/contracts/ERC5496.sol)

```

Comment:
```suggestion
```
Required: no external links
Recommended: It is implied that what follows is the reference impl, so no need to restate it